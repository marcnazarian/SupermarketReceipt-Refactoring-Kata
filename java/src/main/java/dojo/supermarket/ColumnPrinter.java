package dojo.supermarket;

public class ColumnPrinter {
    private int columns;

    public ColumnPrinter(int columns) {

        this.columns = columns;
    }

    String getWhitespace(int whitespaceSize) {
        StringBuilder whitespace = new StringBuilder();
        for (int i = 0; i < whitespaceSize; i++) {
            whitespace.append(" ");
        }
        return whitespace.toString();
    }

    String formatColumns(String leftColumn, String rightColumn) {
        String whitespaces = getWhitespace(this.columns - leftColumn.length() - rightColumn.length());
        return leftColumn + whitespaces + rightColumn;
    }
}
