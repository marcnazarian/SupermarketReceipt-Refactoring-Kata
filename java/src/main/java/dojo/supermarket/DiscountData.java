package dojo.supermarket;

import dojo.supermarket.model.Discount;

public class DiscountData {
    private Discount discount;
    private String productPresentation;
    private String description;
    private double discountAmount;

    public DiscountData(String productPresentation, String description, double discountAmount) {
        this.productPresentation = productPresentation;
        this.description = description;
        this.discountAmount = discountAmount;
    }

    public String getProductPresentation() {
        return productPresentation;
    }

    public String getDescription() {
        return description;
    }

    public double getDiscountAmount() {
        return discountAmount;
    }

}
