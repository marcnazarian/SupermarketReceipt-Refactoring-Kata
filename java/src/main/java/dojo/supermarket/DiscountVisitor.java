package dojo.supermarket;

import dojo.supermarket.model.Product;

public class DiscountVisitor {
    public void visitProductname(Product product) {

    }

    public void visitDescription(String description) {

    }

    public void visitAmount(double discountAmount) {

    }
}
