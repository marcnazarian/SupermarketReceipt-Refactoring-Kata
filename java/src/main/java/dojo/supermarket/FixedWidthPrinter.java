package dojo.supermarket;

public class FixedWidthPrinter {
    private int columns;

    public FixedWidthPrinter(int columns) {

        this.columns = columns;
    }

    private static String getWhitespace(int whitespaceSize) {
        StringBuilder whitespace = new StringBuilder();
        for (int i = 0; i < whitespaceSize; i++) {
            whitespace.append(" ");
        }
        return whitespace.toString();
    }

    public String formatColumns(String leftColumn, String rightColumn) {
        String whitespaces = getWhitespace(this.columns - leftColumn.length() - rightColumn.length());
        return leftColumn +
                whitespaces + rightColumn;
    }
}
