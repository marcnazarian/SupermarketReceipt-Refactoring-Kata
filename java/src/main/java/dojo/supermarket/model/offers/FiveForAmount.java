package dojo.supermarket.model.offers;

import dojo.supermarket.model.Product;

public class FiveForAmount extends NumberOfItemsForAmount {

    public FiveForAmount(Product product, double amount) {
        super(product, 5, amount);
    }


}
