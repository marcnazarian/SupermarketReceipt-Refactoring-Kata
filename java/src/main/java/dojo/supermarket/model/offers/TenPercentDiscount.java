package dojo.supermarket.model.offers;

import dojo.supermarket.model.Product;

public class TenPercentDiscount extends PercentDiscount {
    public TenPercentDiscount(Product product) {
        super(product, 10.0);
    }
}
