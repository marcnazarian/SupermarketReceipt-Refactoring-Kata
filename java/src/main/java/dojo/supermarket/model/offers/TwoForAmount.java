package dojo.supermarket.model.offers;

import dojo.supermarket.model.Discount;
import dojo.supermarket.model.Product;

public class TwoForAmount extends Offer {

    public TwoForAmount(Product product, double amount) {
        super(product, amount);
    }

    @Override
    public Discount getDiscount(Product p, double quantity, double unitPrice, int quantityAsInt) {
        int x = 2;
        if (quantityAsInt < 2) {
            return null;
        }

        double total = argument * quantityAsInt / x + quantityAsInt % 2 * unitPrice;
        double discountN = unitPrice * quantity - total;
        return new Discount(p, "2 for " + argument, discountN);
    }
}
