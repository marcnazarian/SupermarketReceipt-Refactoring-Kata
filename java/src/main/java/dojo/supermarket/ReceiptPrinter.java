package dojo.supermarket;

import dojo.supermarket.model.*;

public class ReceiptPrinter {

    private final int columns;

    public ReceiptPrinter(int columns) {
        this.columns = columns;
    }

    public String printReceipt(Receipt receipt) {
        StringBuilder result = new StringBuilder();
        for (ReceiptItem item : receipt.getItems()) {
            String line = receiptItemLine(item);
            result.append(line);
        }
        for (Discount discount : receipt.getDiscounts()) {
            String discountline = discountLine(discount);
            result.append(discountline + "\n");
        }
        result.append("\n");
        String line = receiptTotalSection(receipt);
        result.append(line);
        return result.toString();
    }

    private String discountLine(Discount discount) {
        DiscountVisitor discountVisitor = new DiscountVisitor();
        discountVisitor.visitProductname(discount.getProduct());
        discountVisitor.visitDescription(discount.getDescription());
        discountVisitor.visitAmount(discount.getDiscountAmount());

        String productPresentation = discount.getProduct().getName();
        String pricePresentation = String.format("%.2f", discount.getDiscountAmount());
        String description = discount.getDescription();
        String leftColumn = description + "(" + productPresentation + ")";
        String rightColumn = "-" + pricePresentation;
        return formatColumns(leftColumn, rightColumn);
    }

    private String formatColumns(String leftColumn, String rightColumn) {
        String whitespaces = getWhitespace(this.columns - leftColumn.length() - rightColumn.length());
        return leftColumn +
                whitespaces + rightColumn;
    }

    private String receiptTotalSection(Receipt receipt) {
        String pricePresentation = String.format("%.2f", (double) receipt.getTotalPrice());
        String total = "Total: ";
        return formatColumns(total, pricePresentation);
    }

    private String receiptItemLine(ReceiptItem item) {

        String price = String.format("%.2f", item.getTotalPrice());
        String name = item.getProduct().getName();

        String line1 = formatColumns(name, price);

        String line = line1 + "\n";

        if (item.getQuantity() != 1) {
            String quantity = presentQuantity(item);
            String unitPrice = String.format("%.2f", item.getPrice());
            line += "  " + unitPrice + " * " + quantity + "\n";
        }
        return line;
    }

    private static String presentQuantity(ReceiptItem item) {
        return ProductUnit.Each.equals(item.getProduct().getUnit())
                ? String.format("%x", (int) item.getQuantity())
                : String.format("%.3f", item.getQuantity());
    }

    private static String getWhitespace(int whitespaceSize) {
        StringBuilder whitespace = new StringBuilder();
        for (int i = 0; i < whitespaceSize; i++) {
            whitespace.append(" ");
        }
        return whitespace.toString();
    }
}
