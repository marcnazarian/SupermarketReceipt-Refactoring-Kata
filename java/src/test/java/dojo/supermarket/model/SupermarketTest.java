package dojo.supermarket.model;

import dojo.supermarket.ReceiptPrinter;
import dojo.supermarket.model.offers.FiveForAmount;
import dojo.supermarket.model.offers.TenPercentDiscount;
import dojo.supermarket.model.offers.TwoForAmount;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class SupermarketTest {

    private SupermarketCatalog catalog;
    private Product toothbrush;
    private Product apples;
    private Product rice;
    private Product toothpaste;
    private Product cherry_tomatoes;
    private ShoppingCart cart;
    private Teller teller;

    @BeforeEach
    private void setUp() {
        catalog = new FakeCatalog();

        toothbrush = new Product("toothbrush", ProductUnit.Each);
        catalog.addProduct(toothbrush, 0.99);

        apples = new Product("apples", ProductUnit.Kilo);
        catalog.addProduct(apples, 1.99);

        rice = new Product("rice", ProductUnit.Each);
        catalog.addProduct(rice, 2.49);

        toothpaste = new Product("toothpaste", ProductUnit.Each);
        catalog.addProduct(toothpaste, 1.79);

        cherry_tomatoes = new Product("cherry tomatoes", ProductUnit.Each);
        catalog.addProduct(cherry_tomatoes, 0.69);

        cart = new ShoppingCart();
        teller = new Teller(catalog);
    }

    @Test
    void no_special_offer() {
        cart.addItemQuantity(apples, 2.5);
        teller.addSpecialOffer(new TenPercentDiscount(toothbrush, 10.0));

        Receipt receipt = teller.checksOutArticlesFrom(cart);

        Approvals.verify(new ReceiptPrinter().printReceipt(receipt));
    }

    @Test
    void buy_two_toothbrushes_get_one_free_normal_toothbrush_price_is_0_99() {
        cart.addItemQuantity(toothbrush, 2);
        teller.addSpecialOffer(new TwoForAmount(toothbrush, 0.99));

        Receipt receipt = teller.checksOutArticlesFrom(cart);

        Approvals.verify(new ReceiptPrinter().printReceipt(receipt));
    }

    @Test
    void discount_20_percent_on_apples_normal_price_is_1_99_per_kilo() {
        cart.addItemQuantity(apples, 2.5);
        teller.addSpecialOffer(new TenPercentDiscount(apples, 20.0));

        Receipt receipt = teller.checksOutArticlesFrom(cart);

        Approvals.verify(new ReceiptPrinter().printReceipt(receipt));
    }

    @Test
    void discount_10_percent_on_rice_normal_price_is_2_49_per_bag() {
        cart.addItemQuantity(rice, 3);
        teller.addSpecialOffer(new TenPercentDiscount(rice, 10.0));

        Receipt receipt = teller.checksOutArticlesFrom(cart);

        Approvals.verify(new ReceiptPrinter().printReceipt(receipt));
    }

    @Test
    void five_tubes_of_toothpaste_for_7_49_normal_price_is_1_79_per_tube() {
        cart.addItemQuantity(toothpaste, 5);
        teller.addSpecialOffer(new FiveForAmount(toothpaste, 7.49));

        Receipt receipt = teller.checksOutArticlesFrom(cart);

        Approvals.verify(new ReceiptPrinter().printReceipt(receipt));
    }

    @Test
    void two_boxes_of_cherry_tomatoes_for_0_99_normal_price_0_69() {
        cart.addItemQuantity(cherry_tomatoes, 2);
        teller.addSpecialOffer(new TwoForAmount(cherry_tomatoes, 0.99));

        Receipt receipt = teller.checksOutArticlesFrom(cart);

        Approvals.verify(new ReceiptPrinter().printReceipt(receipt));
    }

}
