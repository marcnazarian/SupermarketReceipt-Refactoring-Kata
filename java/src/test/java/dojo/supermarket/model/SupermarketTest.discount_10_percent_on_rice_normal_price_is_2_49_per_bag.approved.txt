rice                                7.47
  2.49 * 3
10.0% off(rice)                    -0.75

Total:                              6.72